js/src/Ice/LoggerMiddleware.js

```diff
@@ -15,62 +16,100 @@ export class LoggerMiddleware {
+                    this.logDispatchFailed(response.exceptionDetails, request.current);
                     break;
 
                 default:
-                    this.warning(response.exceptionDetails, request.current);
+                    if (this._traceLevel > 0 || this._warningLevel > 1) {
+                        this.logDispatchFailed(response.exceptionDetails, request.current);
+                    }
                     break;
             }
             return response;
         } catch (ex) {
             if (ex instanceof UserException) {
-                // No warning
-            } else if (ex instanceof RequestFailedException) {
-                if (this._warningLevel > 1) {
-                    this.warning(`${ex}\n${ex.stack}`, request.current);
+                if (this._traceLevel > 0) {
+                    logDispatch(ReplyStatus.UserException, request.current);
+                }
+            } else if (ex instanceof UnknownException) {
+                // always log when middleware installed
+                this.logDispatchFailed(`${ex}\n${ex.stack}`, request.current);
+            } else if (ex instanceof DispatchException) {
+                if (this._traceLevel > 0 || this._warningLevel > 1) {
+                    this.logDispatchFailed(`${ex}\n${ex.stack}`, request.current);
                 }
             } else {
-                this.warning(`${ex}\n${ex.stack}`, request.current);
+                this.logDispatchFailed(`${ex}\n${ex.stack}`, request.current);
             }
             throw ex;
         }
     }
 
-    constructor(next, logger, warningLevel, toStringMode) {
+    constructor(next, logger, traceLevel, traceCat, warningLevel, toStringMode) {
         this._next = next;
         this._logger = logger;
+        this._traceLevel = traceLevel;
+        this._traceCat = traceCat;
         this._warningLevel = warningLevel;
         this._toStringMode = toStringMode;
+
+        DEV: console.assert(traceLevel > 0 || warningLevel > 0);
+    }
+
+    logDispatch(replyStatus, current) {
+        let output = `dispatch of ${current.operation} to `;
+        this.printTarget(output, current);
+        output += ` returned a response with reply status ${replyStatus.name}`;
+        this._logger.trace(this._traceCat, output);
+    }
+
+    logDispatchFailed(exceptionMessage, current) {
+        let output = `failed to dispatch ${current.operation} to `;
+        this.printTarget(output, current);
+
+        if (exceptionMessage !== null) {
+            output += `:\n${exceptionMessage}`;
+        }
+        this._logger.warning(output);
     }
 
-    warning(exceptionMessage, current) {
-        let output = "dispatch exception:";
-        output += `\nidentity: ${identityToString(current.id, this._toStringMode)}`;
-        output += `\nfacet: ${StringUtil.escapeString(current.facet, "", this._toStringMode)}`;
-        output += `\noperation: ${current.operation}`;
+    printTarget(output, current) {
```

Comment:
I would make printTarget a standalone function, so it remains private to this file.

Reply:
Fixed and reworked/tested the code, which was bogus.